### PR TITLE
[Inductor] Stabilize checkpointed reduction autotuning

### LINKED ISSUE #186572: [Inductor] Checkpointed reductions can choose different configs between forward and recompute
### 🐛 Describe the bug

Activation-checkpointed reductions compiled by Inductor can pick different reduction/autotuning configs for the original forward and the checkpoint recompute. When the reduction feeds a discrete downstream decision, low-bit drift between the original forward and recompute can change routing/count logic even though the model code is deterministic.

The important part is that the original forward graph and the AC recompute graph are compiled independently. The recompute graph can materialize backward-needed outputs and therefore have different register pressure. With dynamic R-block selection enabled, that can select a different reduction tree than the original forward.

A minimal repro is below. It uses only PyTorch APIs: `torch.compile`, non-reentrant `checkpoint`, an RMSNorm-like reduction, and a custom compare op that records the original forward output and compares it with the recompute output. No randomness or nondeterminism is injected after the input is created.

```python
import os

import torch
from torch.utils.checkpoint import checkpoint


torch.manual_seed(0)
torch.set_default_device("cuda")

seen = []


@torch.library.custom_op("ac_repro::compare", mutates_args=())
def compare(x: torch.Tensor) -> torch.Tensor:
    # First real call is the original forward. The second real call is the
    # checkpoint recompute during backward.
    if len(seen) == 0:
        seen.append(x.detach().clone())
    else:
        ref = seen[0]
        if not torch.equal(ref, x):
            diff = (ref.float() - x.float()).abs()
            idx = diff.argmax()
            raise RuntimeError(
                "checkpoint recompute mismatch: "
                f"max_abs={diff.max().item()} "
                f"ref={ref.flatten()[idx].item()} "
                f"got={x.flatten()[idx].item()} "
                f"dynamic_rblock={os.environ.get('TORCHINDUCTOR_DYNAMIC_SCALE_RBLOCK', '<default>')}"
            )
    return x


@compare.register_fake
def _(x):
    return torch.empty_like(x)


def setup_context(ctx, inputs, output):
    pass


def backward(ctx, grad):
    return grad


compare.register_autograd(backward, setup_context=setup_context)


def rmsnorm_like(x):
    # Reduction whose exact low bits can depend on the selected reduction tree.
    scale = torch.rsqrt((x.float() * x.float()).mean(dim=-1, keepdim=True) + 1e-6)
    y = x * scale.to(x.dtype)
    return torch.ops.ac_repro.compare.default(y)


def fn(x):
    y = checkpoint(rmsnorm_like, x, use_reentrant=False)
    return y.square().sum()


x = torch.randn(2048, 4096, dtype=torch.bfloat16, requires_grad=True)
compiled = torch.compile(fn, fullgraph=True)
loss = compiled(x)
loss.backward()
print("passed")
```

On an affected Blackwell/GB200-style backend, this fails before any router/topk/count logic is involved, with the compare op reporting a bf16 mismatch between the original forward and checkpoint recompute output.

A workaround is to disable dynamic R-block selection:

```bash
TORCHINDUCTOR_DYNAMIC_SCALE_RBLOCK=0 python repro.py
```

With that setting, the repro passes in the same environment.

This was originally noticed in TorchTitan full activation checkpointing, where the low-bit reduction drift later surfaced as a downstream `repeat_interleave(output_size=...)` mismatch in MoE routing/count handling: https://github.com/pytorch/torchtitan/issues/3336

### Versions

Observed on a recent PyTorch nightly/source build with CUDA on an affected Blackwell/GB200-style backend. The issue appears tied to Inductor dynamic R-block autotuning for independently compiled checkpoint forward/recompute reductions rather than Python-level nondeterminism.

cc @soulitzer @mruberry @kurtamohler @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #186573

Activation checkpointing wants the original forward AC region and its recompute to use the same numerics-affecting compiled strategy. The ideal long-term design is compile-once/reuse for AC subgraphs in AOTAutograd or partitioning, but today those copies are compiled independently.

That independent compilation can change the compiled strategy for reductions. In the failure this fixes, recompute materialized backward-needed outputs, which changed register pressure and let dynamic R-block selection choose a different reduction tree. Low-bit output drift from that difference can then affect downstream discrete decisions such as routing or counts.

This PR keeps the fix Inductor-local and narrower than global deterministic algorithms: use existing AC metadata to put AC-origin reductions on a canonical reduction/autotune policy. The policy uses filtered reduction configs, disables dynamic R-block scaling, disables reduction coordinate-descent and combo retuning choices that can alter the reduction tree, and aligns autotune input-preservation behavior. Unrelated non-AC reductions keep the existing fast path.

The change was authored with assistance from an AI assistant.

Fixes #186572

Test Plan:

```bash
python -m py_compile torch/_inductor/codegen/triton.py torch/_inductor/codegen/triton_combo_kernel.py torch/_inductor/runtime/triton_heuristics.py test/inductor/test_combo_kernels.py test/inductor/test_online_softmax.py
```

```bash
python test/inductor/test_combo_kernels.py ComboKernelMetadataTests.test_checkpointed_standalone_reduction_uses_ac_stable_policy ComboKernelMetadataTests.test_checkpointed_combo_reduction_uses_ac_stable_policy ComboKernelMetadataTests.test_combo_subkernel_fingerprint_includes_ac_stable_policy
```

```bash
python test/inductor/test_online_softmax.py -k test_checkpointed_softmax_uses_ac_stable_reduction_policy
```

```bash
python test/inductor/test_combo_kernels.py -k combo_kernel_coordesc_tunes_largest_subkernel_first
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by xmfan** [comment]
This is the whack-a-mole approach, @eellison @soulitzer I wonder if there's a more general way to have AC regions have same inductor numerics between forward and backward. Are there other inductor features to worry about besides reductions?

**Review by eellison — COMMENTED** [review]
The generic version of this is :
- keep AC region as subgraph in forward/backward
- Use deterministic algorithms / be careful of fusions that would change reduction order. 

We had discussions about this earlier with @soulitzer. unfortunately those conversations are lost to workchat migration. 


I'm worried this implementation is going to be fragile.

edit: just saw @xmfan comment. agreed.

**Comment by sanketpurandare** [comment]
@xmfan @eellison @soulitzer 

  Practically, I can either:

  1. keep this PR as a narrow hotfix and explicitly say it is scoped, or
  2. convert it into a smaller “disable dynamic R-block for AC reductions” mitigation while opening a
     follow-up design issue for reusable AC-region numerics.
 What's your suggestion?

**Comment by eellison** [comment]
@sanketpurandare what do you think about metadata tagging to turn on deterministic algorithms, and turn off dynamic rscale ? Is that sufficient ?

**Comment by sanketpurandare** [comment]
Thanks @xmfan and @eellison. I agree with the concern: this PR addresses the concrete reduction drift we found, but it is not the general AC numerics invariant by itself.


The design I think we should aim for is to make AC regions carry explicit metadata through compile/lowering, then have Inductor apply a centralized “AC numerics-stable” policy for numerics-sensitive kernels instead of adding one-off checks at each symptom. For the failure here, that policy needs at least filtered/deterministic reduction configs and dynamic R-block disabled for AC-origin reductions. The combo-kernel fingerprinting and `optimize_mem` handling are also part of making that stable, since otherwise forward and recompute can still benchmark/autotune under different conditions.

This is what Codex suggests:

> I would be careful about phrasing this as globally turning on `torch.use_deterministic_algorithms(True)` semantics. What we need here is an Inductor-local policy for AC-tagged regions: preserve or stabilize numerics-affecting lowering/autotune choices between the original forward and recompute. Longer term, keeping the AC region as an explicit subgraph in forward/backward and sharing or normalizing those choices is the more principled version.
> 
> So my read is: metadata tagging + deterministic reduction config filtering + disabling dynamic R-block is sufficient for the observed reduction-tree bug, but not a universal guarantee that arbitrary AC recompute is bitwise equal to forward. I can rework this PR toward that metadata-policy, or keep this as a scoped hotfix and open a follow-up issue/PR for the broader AC numerics-equivalence design.

**Comment by soulitzer** [comment]
From discussions with @eellison and others a while back, the idea was to try to make sure inductor only compiles each region of the graph once, and if you had duplicated nodes from AC/SAC you would partition those parts of the graph out, compile it only once and reuse that for both during the forward and backward. 

If there's a simpler way that avoids this involved graph surgery in AOTAutograd, that sounds very worth exploring as well though.

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -3163,15 +3163,19 @@ def create_constant_mask(self, entry: IterationRangesRoot) -> str:
...
+
     @staticmethod
     def _has_stride1_on_rdim(index) -> bool:
-        # These analysis is only needed in deterministic mode so far
-        # to filter triton configs. Return false immediately to avoid
-        # increasing compilation time when the mode is off.
-        if not (
-            config.deterministic or config.test_configs.force_filter_reduction_configs
-        ):
-            return False
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:3174]: I dont understand why we have these lines to begin with cc @shunting314

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:3174]: The flag was only needed during deterministic mode. Skip the analysis to save compilation time.

I can imagine removing these check will increase compilation time a bit when deterministic-mode is off.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -6443,6 +6484,11 @@ def add_constexpr_arg(arg_name):
         # every new node will increase the peak memory and our greedy approach would
         # introduce a lot of unnecessary cpu copies.
         optimize_mem = V.graph.is_inference or V.graph.is_backward
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:6486]: Can you say why we needed this? for cache hit, or, just to paranoid be sure we have same configs ? bc this just says to move inputs to cpu as needed to avoid duplicate copies. if its used as a cache key, we shoudl fix that (not in this pr).

File: torch/_inductor/codegen/triton_combo_kernel.py [hunk 3]
```diff
@@ -774,14 +784,18 @@ def jit_line(
...
             # graphs must keep it to preserve benchmark inputs across retries.
-            "optimize_mem": V.graph.is_inference or V.graph.is_backward,
+            "optimize_mem": optimize_mem,
             **self.triton_kernel_cls.inductor_meta_common(),
         }
+        if has_ac_stable_reduction:
+            inductor_meta["ac_stable_reduction"] = True
+            inductor_meta["dynamic_scale_rblock"] = False
+            inductor_meta["disable_reduction_coordesc"] = True
+            inductor_meta["disable_combo_kernel_autotune"] = True
```
Comment (eellison) [inline torch/_inductor/codegen/triton_combo_kernel.py:794]: Hmm, so, we dont have any way of guaranteeing the forward wont be in combo kernel , do we ?

File: torch/_inductor/runtime/triton_heuristics.py [hunk 4]
```diff
@@ -1897,12 +1897,17 @@ def _should_coordesc_tune(self) -> bool:
...
             return False
-        # Deterministic mode forbids tuning RBLOCK / num_warps for reductions
-        # because those knobs shift numerics.
-        if self.deterministic_mode and self.heuristic_type in (
+        reduction_heuristic = self.heuristic_type in (
             HeuristicType.REDUCTION,
             HeuristicType.PERSISTENT_REDUCTION,
             HeuristicType.SPLIT_SCAN,
+        )
+        # These knobs can shift reduction numerics by changing the tree.
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:1905]: cc @shunting314  to review

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -6257,6 +6266,38 @@ def inductor_meta_per_kernel(self) -> dict[str, Any]:
...
         return out
 
+    @cache_on_self
+    def _has_ac_recompute_origins(self) -> bool:
+        from torch.utils.checkpoint import CheckpointPolicy
+
+        for scheduler_node in self.features.scheduler_nodes():
+            node = scheduler_node.node
+            if node is None:
+                continue
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:6276]: I really dont like using origins as a correctness measure. What this is doing is making origins load bearing for correctness, which is not the case righ tnow. 

So it's basically taking a huge dependency on something that no one has validated. arguably, the same thing could be said for meta, which is a related issue cc @zou3519

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -6257,6 +6266,38 @@ def inductor_meta_per_kernel(self) -> dict[str, Any]:
                 out["kernel_flop"] = flops
         return out
 
+    @cache_on_self
+    def _has_ac_recompute_origins(self) -> bool:
+        from torch.utils.checkpoint import CheckpointPolicy
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:6271]: I wonder if we should set this in torchtitan, instead of on by default ? today we dont have any guarantee and it might interfere/hurt with perf.. im not sure

File: torch/_inductor/codegen/triton_combo_kernel.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/codegen/triton_combo_kernel.py:794]: In this case, what you need is meta on both the fwd and backward. although the more complete fix is just subgraphs.

**Review by eellison — COMMENTED** [review]
We can accept to unblock, although this does need to be done in a more systematic manner. cc @shunting314 to review as well.
